src/back/spv/writer.rs

```diff
@@ -450,147 +838,92 @@ impl Writer {
-                            instruction.add_operand(low);
-                            instruction.add_operand(high);
+                            vec![low, high]
                         }
                         _ => unreachable!(),
                     },
                     _ => unreachable!(),
-                }
+                };
 
+                let type_id = self.get_type_id(arena, constant.ty);
+                let instruction = self.instruction_constant(type_id, id, values);
                 (instruction, id)
             }
             crate::ConstantInner::Uint(val) => {
-                let type_id = self.get_type_id(arena, constant.ty);
-
-                let mut instruction = Instruction::new(Op::Constant);
-                instruction.set_type(type_id);
-                instruction.set_result(id);
-
                 let ty = &ir_module.types[constant.ty];
-                match ty.inner {
+                let values = match ty.inner {
                     crate::TypeInner::Scalar { kind: _, width } => match width {
-                        32 => {
-                            instruction.add_operand(val as u32);
-                        }
+                        32 => vec![val as u32],
                         64 => {
                             let (low, high) = ((val >> 32) as u32, val as u32);
-                            instruction.add_operand(low);
-                            instruction.add_operand(high);
+                            vec![low, high]
                         }
                         _ => unreachable!(),
                     },
                     _ => unreachable!(),
-                }
+                };
 
+                let type_id = self.get_type_id(arena, constant.ty);
+                let instruction = self.instruction_constant(type_id, id, values);
                 (instruction, id)
             }
             crate::ConstantInner::Float(val) => {
-                let type_id = self.get_type_id(arena, constant.ty);
-
-                let mut instruction = Instruction::new(Op::Constant);
-                instruction.set_type(type_id);
-                instruction.set_result(id);
-
                 let ty = &ir_module.types[constant.ty];
-                match ty.inner {
+                let values = match ty.inner {
                     crate::TypeInner::Scalar { kind: _, width } => match width {
-                        32 => {
-                            instruction.add_operand((val as f32).to_bits());
-                        }
+                        32 => vec![(val as f32).to_bits()],
                         64 => {
                             let bits = f64::to_bits(val);
                             let (low, high) = ((bits >> 32) as u32, bits as u32);
-                            instruction.add_operand(low);
-                            instruction.add_operand(high);
+                            vec![low, high]
```

Comment:
same here and in other places - let's be cautious about heap allocations

Reply:
Yes thanks for keep telling me that, I still tend to use vectors everywhere, but I am learning it 😄